velox/substrait/VeloxToSubstraitPlan.cpp

```diff
@@ -170,13 +176,119 @@ void VeloxToSubstraitPlanConvertor::toSubstrait(
+  // 2. aggregateMasksSize != 0, aggregatesSize = aggregateMasksSize.
+  auto aggregates = aggregateNode->aggregates();
+  auto aggregateMasks = aggregateNode->aggregateMasks();
+  int64_t aggregatesSize = aggregates.size();
+  int64_t aggregateMasksSize = aggregateMasks.size();
+  VELOX_CHECK_GE(aggregatesSize, aggregateMasksSize);
+
+  for (int64_t i = 0; i < aggregatesSize; i++) {
+    ::substrait::AggregateRel_Measure* aggMeasures =
+        aggregateRel->add_measures();
+
+    auto aggMaskExpr = aggregateMasks.at(i);
+    // Set substrait filter.
+    ::substrait::Expression* aggFilter = aggMeasures->mutable_filter();
+    if (aggMaskExpr.get()) {
+      aggFilter->MergeFrom(exprConvertor_->toSubstraitExpr(
+          arena,
+          std::dynamic_pointer_cast<const core::ITypedExpr>(aggMaskExpr),
+          inputType));
+    } else {
+      // Set null.
+      aggFilter = NULL;
+    }
+
+    // Process measure, eg:sum(a).
+    const auto& aggregatesExpr = aggregates.at(i);
+    ::substrait::AggregateFunction* aggFunction =
+        aggMeasures->mutable_measure();
+
+    // Aggregation function name.
+    const auto& funName = aggregatesExpr->name();
+    // set aggFunction args.
+    const auto& aggregatesExprInputs = aggregatesExpr->inputs();
+
+    int64_t aggregatesExprInputsSize = aggregatesExprInputs.size();
+    for (int64_t j = 0; j < aggregatesExprInputsSize; j++) {
+      ::substrait::Expression* aggFunctionExpr = aggFunction->add_args();
+      core::TypedExprPtr expr = aggregatesExprInputs[j];
+
+      // If the expr is CallTypedExpr, people need do project firstly.
+      if (auto aggregatesExprInput =
+              std::dynamic_pointer_cast<const core::CallTypedExpr>(expr)) {
+        VELOX_NYI("In Velox Plan, the aggregates type cannot be CallTypedExpr");
+      } else if (
+          auto aggregatesExprInput =
+              std::dynamic_pointer_cast<const core::ConstantTypedExpr>(expr) &&
+              funName == "count") {
+        // For case count(1).
+        aggFunction->clear_args();
+      } else {
+        // FieldAccessTypedExpr and Cast.
+        aggFunctionExpr->MergeFrom(
+            exprConvertor_->toSubstraitExpr(arena, expr, inputType));
+      }
+    }
+
+    // Set substrait aggregate Function reference and output type.
+    aggFunction->set_function_reference(functionMap_[funName]);
+
+    aggFunction->mutable_output_type()->MergeFrom(
+        typeConvertor_->toSubstraitType(arena, aggregatesExpr->type()));
+
+    // Set substrait aggregate Function phase.
+    aggFunction->MergeFrom(setAggregateSteps(arena, aggregateNode));
+  }
+
+  // Direct output.
+  aggregateRel->mutable_common()->mutable_direct();
+}
+
 void VeloxToSubstraitPlanConvertor::constructFunctionMap() {
   // TODO: Fetch all functions from velox's registry.
 
   functionMap_["plus"] = 0;
   functionMap_["multiply"] = 1;
   functionMap_["lt"] = 2;
   functionMap_["divide"] = 3;
+  functionMap_["count"] = 4;
```

Comment:
Let's add proper logic to lookup functions in the registry in a follow-up PR.

Reply:
Sure.